cmd/snap-bootstrap/partition/sfdisk.go

```diff
@@ -102,15 +102,28 @@ func (dl *DeviceLayout) CreateMissing(pv *gadget.LaidOutVolume) ([]DeviceStructu
 	buf, created := buildPartitionList(dl.partitionTable, pv)
 
 	// Write the partition table, note that sfdisk will re-read the
-	// partition table by itself: see disk-utils/sfdisk.c:write_changes()
-	cmd := exec.Command("sfdisk", dl.Device)
+	// partition table by itself: see disk-utils/sfdisk.c:write_changes().
+	// Use --force, we have existing target device partitions mounted.
+	cmd := exec.Command("sfdisk", "--force", dl.Device)
 	cmd.Stdin = buf
 	if output, err := cmd.CombinedOutput(); err != nil {
 		return created, osutil.OutputErr(output, err)
 	}
 
 	// Add an extra delay to prevent reads to take place immediately
-	time.Sleep(250 * time.Millisecond)
+	time.Sleep(100 * time.Millisecond)
+
+	// XXX: force partition table re-read and create device nodes. Ideally
+	// we should trigger node creation directly instead of relying on a
+	// side effect of using a specific utility.
+	// Note: blockdev --rereadpt doesn't work, partx must be used
+	output, err := exec.Command("partx", "-u", dl.Device).CombinedOutput()
+	if err != nil {
+		return nil, osutil.OutputErr(output, err)
+	}
+
+	// Add an extra delay to prevent reads to take place immediately
```

Comment:
What is the origin of the raciness? Are we waiting for udev to create device nodes? Is this something else that we know about?

Reply:
The use case scenario for `--force` is the following: the partition table is read from with `sfdisk --dump`, new partitions from the gadget are added, and the partition table is updated. Existing partitions are not modified. Partition consistency checking is performed for the volume layout when creating a `LaidOutVolume` and we also verify if the supplied gadget layout matches the existing disk (for the partitions that are already there). At the point `snap-bootstrap create-partitions` runs, the system is running in ephemeral mode and is seeded from the recovery partition. I'll check the sfdisk source code to see if there's any other way around this, but I'm not sure if there's much we can do in this case.

Reply:
Regarding the partition table re-read delays, the race seems to be internal to the kernel and its need was determined empirically. The comment on the corresponding delay in the sfdisk source code states: "Let's wait a little bit. It's possible that our system is still busy with a previous re-read ioctl (on sfdisk start) or with another task related to the write to the device."